# Markov chains

A Markov chain is a stochastic process $a_1$, $a_2$, ..., $a_n$, ... where the probability distribution of having a state $a_{n+1}$ at the $(n+1)$th step only depends on the $n$th state $a_n$. i.e, a Markov process is fully defined by the probabilities $p(a_{n+1} | a_n)$.

For example, let us consider two coins: one weighted and one fair. The weighted coin has a 0.6 probability of getting heads. If we get heads on the $n$th coin toss, we use the fair coin in the next round but use the weighted coin otherwise. The probabilities can be summarized as follows:

| Transition | Probability|
|------------|------------|
| H -> H     | 0.5       |
| H -> T     | 0.5       |
| T -> H     | 0.6       |
| T -> T     | 0.4       |

We can simulate this as follows:

In [16]:
import numpy as np

def simulate_coin_toss_game(start_heads, nsteps):
    result = []
    heads = start_heads
    for i in range(nsteps):
        if heads:
            # fair coin toss
            heads = np.random.choice([True, False])
        else:
            # weighted coin with 60% heads
            heads = np.random.choice([True, False], p=[0.6, 0.4])
        result.append(heads)

    return result

result = simulate_coin_toss_game(False, 100)
print([["H", "T"][int(r)] for r in result])
print(f"% heads = {sum(result) / len(result)}")

['T', 'T', 'H', 'T', 'T', 'T', 'H', 'H', 'H', 'H', 'H', 'T', 'T', 'T', 'H', 'T', 'T', 'H', 'H', 'T', 'T', 'T', 'H', 'H', 'H', 'T', 'H', 'T', 'T', 'H', 'H', 'T', 'T', 'T', 'T', 'T', 'H', 'T', 'T', 'T', 'H', 'T', 'H', 'T', 'T', 'H', 'H', 'T', 'H', 'T', 'H', 'T', 'T', 'H', 'T', 'T', 'H', 'T', 'T', 'H', 'H', 'H', 'H', 'T', 'T', 'H', 'H', 'H', 'H', 'H', 'H', 'T', 'T', 'T', 'H', 'T', 'T', 'H', 'T', 'T', 'T', 'T', 'H', 'T', 'H', 'H', 'T', 'H', 'T', 'T', 'H', 'T', 'T', 'H', 'H', 'H', 'H', 'H', 'T', 'H']
% heads = 0.53


Repeating this game, we see that the percentage of heads is on average slightly over 50% most of the time. 

So how does the probability of heads change after each step? Does it converge to anything? The following code computes the probability of heads at each step.

In [17]:
def simulate_coin_toss_game_probability(start_heads, nsteps):
    heads_prob = 0.5 if start_heads else 0.6
    result = [heads_prob]
    for i in range(nsteps):
        # fair coin if H, weighted coin if T
        heads_prob = 0.5 * heads_prob + 0.6 * (1 - heads_prob)
        result.append(heads_prob)

    return result

result = simulate_coin_toss_game_probability(True, 15)
print(result)

result = simulate_coin_toss_game_probability(False, 15)
print(result)

[0.5, 0.55, 0.5449999999999999, 0.5455, 0.54545, 0.545455, 0.5454545, 0.54545455, 0.545454545, 0.5454545454999999, 0.5454545454499999, 0.5454545454549999, 0.5454545454545, 0.54545454545455, 0.545454545454545, 0.5454545454545454]
[0.6, 0.54, 0.546, 0.5454, 0.5454600000000001, 0.545454, 0.5454546, 0.5454545399999999, 0.545454546, 0.5454545453999999, 0.5454545454599999, 0.5454545454540001, 0.5454545454546, 0.54545454545454, 0.545454545454546, 0.5454545454545454]


Regardless of whether we start the game with an H or not, we see that the probability of H converges to the same number after several steps. This is called the steady state probability. The steady state probability can be calculated from the transition probabilities using the property that the probabilities in the $(n+1)$th step are the same as the $n$th step once the steady state is reached. i.e., in this case,

$p_{n+1}[H] = p_{n}[H] = 0.5 p_n[H] + 0.6 p_n[T] = 0.5 p_n[H] + 0.6 (1-p_n[H])$

This has the solution $p_n[H] = 6/11 = 0.54...$, which is the same number we got from the simulation.